src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,218 @@
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IVsUIService<IVsExtensibility, IVsExtensibility3> _extensibility;
+        private readonly IVsOnlineServices _vsOnlineServices;
+        private readonly IUserNotificationServices _userNotificationServices;
+        private readonly IWaitIndicator _waitService;
+        private readonly IRoslynServices _roslynServices;
+        private readonly Workspace _workspace;
+        private Dictionary<(string, string), SemaphoreSlim> _languageServiceDone;
+
+        [ImportingConstructor]
+        public RenamerProjectTreeActionHandler(
+            UnconfiguredProject unconfiguredProject,
+            IUnconfiguredProjectVsServices projectVsServices,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace,
+            IEnvironmentOptions environmentOptions,
+            IUserNotificationServices userNotificationServices,
+            IRoslynServices roslynServices,
+            IWaitIndicator waitService,
+            IVsOnlineServices vsOnlineServices,
+            IProjectThreadingService threadingService,
+            IVsUIService<IVsExtensibility, IVsExtensibility3> extensibility)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _projectVsServices = projectVsServices;
+            _workspace = workspace;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _roslynServices = roslynServices;
+            _waitService = waitService;
+            _vsOnlineServices = vsOnlineServices;
+            _threadingService = threadingService;
+            _extensibility = extensibility;
+            _languageServiceDone = new Dictionary<(string, string), SemaphoreSlim>();
+        }
+
+        public void HandleRename(string oldFilePath, string newFilePath)
+        {
+            string newFileWithExtension = Path.GetFileName(newFilePath);
+            if (_languageServiceDone.TryGetValue((oldFilePath, newFileWithExtension), out SemaphoreSlim semLanguegeServiceDone))
+            {
+                semLanguegeServiceDone.Release();
+            }
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            Requires.NotNull(context, nameof(Context));
+            Requires.NotNull(node, nameof(node));
+            Requires.NotNullOrEmpty(value, nameof(value));
+
+            // These variables are need to synchronize with Roslyn
+            string oldFilePath = node.FilePath;
+            string newFileNameWithExtension = value;
+
+            // Do not offer to rename the file in VS Online scenarios.
+            if (_vsOnlineServices.ConnectedToVSOnline)
```

Comment:
This will break renames of files in VSOnline scenarios because base.RenameAsync is never called.